### ISSUE TITLE
Cannot utilize insertId in any meaningful way, RuntimeException: Database closed

### ISSUE THREAD

**Issue by ajthenewguy** [clue/reactphp-sqlite#41]
I am trying to get the insert ID value for after the insert query, which seems to me like a pretty common practice. This is reporting a RuntimeException: "Database closed."

```php
    public function testTheDatabase()
    {
        $loop = Loop::get();
        $expected = [
            'id' => 1,
            'name' => 'Flounder',
            'color' => 'multi',
            'size' => 7,
            'created_at' => time(),
            'updated_at' => null
        ];

        $this->db = (new \Clue\React\SQLite\Factory(Loop::get()))->openLazy($this->getDatabaseFile()->getPath());

        $this->db->exec('CREATE TABLE IF NOT EXISTS test (
            id INTEGER PRIMARY KEY,
            name VARCHAR (30) NOT NULL,
            color VARCHAR (10) DEFAULT NULL,
            size INTEGER (2) DEFAULT NULL,
            created_at TIMESTAMP DEFAULT (strftime(\'%s\',\'now\')),
            updated_at TIMESTAMP
        )');

        $this->db->query('INSERT INTO test (name, color, size) VALUES (\'Flounder\', \'multi\', 7)');
        $this->db->query("SELECT * FROM test")->then(function ($result) use ($expected) {
            $this->assertTrue(isset($result->rows));
            $this->assertCount(1, $result->rows);
            $this->assertEquals(array_keys($expected), $result->columns);
            $this->assertEquals([$expected], $result->rows);
        })->done();


        $this->db->query('INSERT INTO test (name, color, size) VALUES (\'Sebastian\', \'red\', 6)')->then(function ($Result) {

            $insertId = $Result->insertId;
            print "\nLine: ".__LINE__."\nInserted id: ";
            var_dump($insertId);
            print "\n";

            $this->db->query("SELECT * FROM test WHERE id = ?", [$insertId])->then(function ($result) {

                print "\n" . __LINE__ . "\n";
                // Line: 67
                var_dump($result);
                // Inserted id: int(2)
                print "\n";

            }, function (\Exception $e) {
                print "\n" . __LINE__ . ": " . get_debug_type($e) . ': ' . $e->getMessage() . "\n\n";
                // 80: RuntimeException: Database closed
            })->done();
        })->done();


        $this->db->quit();
        $loop->run();
    }
```

I can't get the query in the "then" callback to work at all, it keeps returning that error.

This is running in a PhpUnit test and the database is created and destroyed each time (note the insert ID of 2).

Here is the PhpUnit output:

```
php -v
PHP 8.0.9 (cli) (built: Jul 29 2021 17:50:26) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.9, Copyright (c) Zend Technologies

./test --filter=DatabaseTest::testTheDatabase
PHPUnit 10.0-dev by Sebastian Bergmann and contributors.

Runtime:       PHP 8.0.9
Configuration: /Users/allen/lib/ajthenewguy/php8-api-server/phpunit.xml

.                                                                   1 / 1 (100%)
Line: 67
Inserted id: int(2)


80: RuntimeException: Database closed



Time: 00:00.159, Memory: 8.00 MB

OK (1 test, 4 assertions)
```

Here is the stack-trace:

```
#0 ~project-root/vendor/clue/reactphp-sqlite/src/Io/ProcessIoDatabase.php(88): Clue\React\SQLite\Io\ProcessIoDatabase->send('query', Array)
#1 ~project-root/vendor/clue/reactphp-sqlite/src/Io/LazyDatabase.php(102): Clue\React\SQLite\Io\ProcessIoDatabase->query('SELECT * FROM t...', Array)
#2 ~project-root/vendor/react/promise/src/FulfilledPromise.php(28): Clue\React\SQLite\Io\LazyDatabase->Clue\React\SQLite\Io\{closure}(Object(Clue\React\SQLite\Io\ProcessIoDatabase))
#3 ~project-root/vendor/react/promise/src/Promise.php(31): React\Promise\FulfilledPromise->then(Object(Closure), NULL, NULL)
#4 ~project-root/vendor/clue/reactphp-sqlite/src/Io/LazyDatabase.php(112): React\Promise\Promise->then(Object(Closure))
#5 ~project-root/tests/Feature/DatabaseTest.php(71): Clue\React\SQLite\Io\LazyDatabase->query('SELECT * FROM t...', Array)
#6 ~project-root/vendor/react/promise/src/FulfilledPromise.php(28): Tests\Feature\DatabaseTest->Tests\Feature\{closure}(Object(Clue\React\SQLite\Result))
#7 ~project-root/vendor/react/promise/src/Promise.php(134): React\Promise\FulfilledPromise->then(Object(Closure), NULL)
#8 ~project-root/vendor/react/promise/src/Promise.php(168): React\Promise\Promise::React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#9 ~project-root/vendor/react/promise/src/Promise.php(231): React\Promise\Promise->settle(Object(React\Promise\FulfilledPromise))
#10 ~project-root/vendor/react/promise/src/FulfilledPromise.php(42): React\Promise\Promise::React\Promise\{closure}(Object(Clue\React\SQLite\Result))
#11 ~project-root/vendor/react/promise/src/Promise.php(66): React\Promise\FulfilledPromise->done(Object(Closure), Object(Closure))
#12 ~project-root/vendor/react/promise/src/Promise.php(168): React\Promise\Promise::React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#13 ~project-root/vendor/react/promise/src/Promise.php(231): React\Promise\Promise->settle(Object(React\Promise\FulfilledPromise))
#14 ~project-root/vendor/react/promise/src/FulfilledPromise.php(42): React\Promise\Promise::React\Promise\{closure}(Object(Clue\React\SQLite\Result))
#15 ~project-root/vendor/react/promise/src/Promise.php(135): React\Promise\FulfilledPromise->done(Object(Closure), Object(Closure), Object(Closure))
#16 ~project-root/vendor/react/promise/src/Promise.php(168): React\Promise\Promise::React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#17 ~project-root/vendor/react/promise/src/Promise.php(231): React\Promise\Promise->settle(Object(React\Promise\FulfilledPromise))
#18 ~project-root/vendor/react/promise/src/FulfilledPromise.php(42): React\Promise\Promise::React\Promise\{closure}(Object(Clue\React\SQLite\Result))
#19 ~project-root/vendor/react/promise/src/Promise.php(135): React\Promise\FulfilledPromise->done(Object(Closure), Object(Closure), Object(Closure))
#20 ~project-root/vendor/react/promise/src/Promise.php(168): React\Promise\Promise::React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#21 ~project-root/vendor/react/promise/src/Promise.php(231): React\Promise\Promise->settle(Object(React\Promise\FulfilledPromise))
#22 ~project-root/vendor/react/promise/src/Deferred.php(36): React\Promise\Promise::React\Promise\{closure}(Array)
#23 ~project-root/vendor/clue/reactphp-sqlite/src/Io/ProcessIoDatabase.php(54): React\Promise\Deferred->resolve(Array)
#24 ~project-root/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(123): Clue\React\SQLite\Io\ProcessIoDatabase->Clue\React\SQLite\Io\{closure}(Array)
#25 ~project-root/vendor/clue/ndjson-react/src/Decoder.php(131): Evenement\EventEmitter->emit('data', Array)
#26 ~project-root/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(123): Clue\React\NDJson\Decoder->handleData(Array)
#27 ~project-root/vendor/react/stream/src/ReadableResourceStream.php(148): Evenement\EventEmitter->emit('data', Array)
#28 ~project-root/vendor/react/event-loop/src/StreamSelectLoop.php(245): React\Stream\ReadableResourceStream->handleData(Resource id #320)
#29 ~project-root/vendor/react/event-loop/src/StreamSelectLoop.php(212): React\EventLoop\StreamSelectLoop->waitForStreamActivity(NULL)
#30 ~project-root/tests/Feature/DatabaseTest.php(88): React\EventLoop\StreamSelectLoop->run()
#31 ~project-root/vendor/phpunit/phpunit/src/Framework/TestCase.php(1347): Tests\Feature\DatabaseTest->testTheDatabase()
#32 ~project-root/vendor/phpunit/phpunit/src/Framework/TestCase.php(805): PHPUnit\Framework\TestCase->runTest()
#33 ~project-root/vendor/phpunit/phpunit/src/Framework/TestRunner.php(114): PHPUnit\Framework\TestCase->runBare()
#34 ~project-root/vendor/phpunit/phpunit/src/Framework/TestCase.php(546): PHPUnit\Framework\TestRunner->run(Object(Tests\Feature\DatabaseTest), Object(PHPUnit\Framework\TestResult))
#35 ~project-root/vendor/phpunit/phpunit/src/Framework/TestSuite.php(511): PHPUnit\Framework\TestCase->run(Object(PHPUnit\Framework\TestResult))
#36 ~project-root/vendor/phpunit/phpunit/src/Framework/TestSuite.php(511): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#37 ~project-root/vendor/phpunit/phpunit/src/Framework/TestSuite.php(511): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#38 ~project-root/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(441): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#39 ~project-root/vendor/phpunit/phpunit/src/TextUI/Application.php(109): PHPUnit\TextUI\TestRunner->run(Object(PHPUnit\Framework\TestSuite))
#40 ~project-root/vendor/phpunit/phpunit/src/TextUI/Application.php(77): PHPUnit\TextUI\Application->run(Array, true)
#41 ~project-root/vendor/phpunit/phpunit/phpunit(84): PHPUnit\TextUI\Application::main()
#42 {main}
```

composer.json excerpt:
```
    "require": {
        "react/event-loop": "^1.2",
        "react/http": "^1.5",
        "react/stream": "^1.2",
        "react/promise": "^2.8",
        "vlucas/phpdotenv": "5.3.x-dev",
        "nesbot/carbon": "^2.51",
        "clue/stdio-react": "^2.4",
        "firebase/php-jwt": "^5.4",
        "clue/reactphp-sqlite": "^1.1",
        "react/mysql": "^0.5.5"
    },
```

**Comment by clue** [comment]
@ajthenewguy Thanks for reporting, this is an interesting one!

This library did exactly what you told it to do, but it looks like you may want to tell it to do something else :-) Each query actually executes asynchronously because it may take a while. Because of how you execute your final `SELECT` only once your `INSERT` is *completed*, but execute your `quit()` right after *queuing* your `INSERT`, you end up executing this sequence:

```php
$this->db->exec('CREATE TABLE IF NOT EXISTS test');
$this->db->query('INSERT INTO test (name, color, size) VALUES (\'Flounder\', \'multi\', 7)');
$this->db->query("SELECT * FROM test");
$this->db->query('INSERT INTO test …');
$this->db->quit();
$this->db->query("SELECT * FROM test WHERE id = …");
```

Hence you're receiving a `Database closed` rejection for your final `SELECT`.

In your case, it looks like you may want to execute the `quit()` only after your final `query()`. For instance, you can invoke the `quit()` call inside your fulfillment handler for the `INSERT` statement. Note that you may want to also include this in a rejection handler in case the `INSERT` statement fails with an error.

I hope this helps :+1:

**Comment by ajthenewguy** [comment]
I see. In the case where the app is simply creating a new record and
returning the ID, if there are no further tasks, wouldn’t the loop end
before completing the select?

Also, should the connection explicitly be closed after the final query
success handler (or in the rejection handler), or does the 60 second
expiration take care of that?

If the connection is not closed and another request is received within 60
seconds is the same connection reused or is a new one created?

Thanks for your help understanding how this works.

**Comment by clue** [comment]
The loop will automatically keep running as long as there are any listeners attached (open streams, timers or signals).

In a long-running application you would normally keep your connection open for any further queries that might reuse the same connection at a later time. If you're using the lazy connection, it will automatically reuse the same connection within its idle setting (keepalive). Once this timer is exceeded, it will soft-close the underlying connection, but will automatically create a new connection for any further queries.

If you're sure you don't need the connection anymore, you may call `quit()` which will wait for all commands to complete and then end the connection.

**Comment by ajthenewguy** [comment]
Thanks so much for your help!
